File: src/nl/rubensten/texifyidea/highlighting/LatexSyntaxHighlighter.java
package nl.rubensten.texifyidea.highlighting;

import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.TokenSet;
import nl.rubensten.texifyidea.LatexLexerAdapter;
import nl.rubensten.texifyidea.psi.LatexTypes;
import org.jetbrains.annotations.NotNull;

/**
 * @author Ruben Schellekens, Sten Wessel
 */
public class LatexSyntaxHighlighter extends SyntaxHighlighterBase {

    /*
     * TextAttributesKeys
     */
    public static final TextAttributesKey BRACES = TextAttributesKey.createTextAttributesKey(
            "LATEX_BRACES",
            DefaultLanguageHighlighterColors.BRACES
    );

    public static final TextAttributesKey BRACKETS = TextAttributesKey.createTextAttributesKey(
            "LATEX_BRACKETS",
            DefaultLanguageHighlighterColors.BRACKETS
    );

    public static final TextAttributesKey OPTIONAL_PARAM = TextAttributesKey.createTextAttributesKey(
            "LATEX_OPTIONAL_PARAM",
            DefaultLanguageHighlighterColors.PARAMETER
    );

    public static final TextAttributesKey COMMAND = TextAttributesKey.createTextAttributesKey(
            "LATEX_COMMAND",
            DefaultLanguageHighlighterColors.KEYWORD
    );

    public static final TextAttributesKey COMMAND_MATH_INLINE = TextAttributesKey.createTextAttributesKey(
            "LATEX_COMMAND_MATH_INLINE",
            DefaultLanguageHighlighterColors.VALID_STRING_ESCAPE
    );

    public static final TextAttributesKey COMMAND_MATH_DISPLAY = TextAttributesKey
            .createTextAttributesKey(
            "LATEX_COMMAND_MATH_DISPLAY",
            DefaultLanguageHighlighterColors.VALID_STRING_ESCAPE
    );

    public static final TextAttributesKey COMMENT = TextAttributesKey.createTextAttributesKey(
            "LATEX_COMMENT",
            DefaultLanguageHighlighterColors.LINE_COMMENT
    );

    public static final TextAttributesKey INLINE_MATH = TextAttributesKey.createTextAttributesKey(
            "LATEX_INLINE_MATH",
            DefaultLanguageHighlighterColors.STRING
    );

    public static final TextAttributesKey DISPLAY_MATH = TextAttributesKey.createTextAttributesKey(
            "LATEX_DISPLAY_MATH",
            DefaultLanguageHighlighterColors.STRING
    );

    public static final TextAttributesKey STAR = TextAttributesKey.createTextAttributesKey(
            "LATEX_STAR",
            DefaultLanguageHighlighterColors.DOT
    );

    public static final TextAttributesKey LABEL_DEFINITION = TextAttributesKey.createTextAttributesKey(
            "LATEX_LABEL_DEFINITION",
            DefaultLanguageHighlighterColors.IDENTIFIER
    );

    public static final TextAttributesKey LABEL_REFERENCE = TextAttributesKey.createTextAttributesKey(
            "LATEX_LABEL_REFERENCE",
            DefaultLanguageHighlighterColors.IDENTIFIER
    );

    /*
     * TokenSets
     */
    private static final TokenSet COMMAND_TOKENS = TokenSet.create(
            LatexTypes.COMMAND_TOKEN,
            LatexTypes.BEGIN_TOKEN,
            LatexTypes.END_TOKEN
    );

    /*
     * TextAttributeKey[]s
     */
    private static final TextAttributesKey[] BRACES_KEYS = new TextAttributesKey[] {
            BRACES
    };

    private static final TextAttributesKey[] BRACKET_KEYS = new TextAttributesKey[] {
            BRACKETS
    };

    private static final TextAttributesKey[] COMMAND_KEYS = new TextAttributesKey[] {
            COMMAND
    };

    private static final TextAttributesKey[] COMMENT_KEYS = new TextAttributesKey[] {
            COMMENT
    };

    private static final TextAttributesKey[] INLINE_MATH_KEYS = new TextAttributesKey[] {
            INLINE_MATH
    };

    private static final TextAttributesKey[] DISPLAY_MATH_KEYS = new TextAttributesKey[] {
            DISPLAY_MATH
    };

    private static final TextAttributesKey[] STAR_KEYS = new TextAttributesKey[] {
            STAR
    };

    private static final TextAttributesKey[] LABEL_DEFINITION_KEYS = new TextAttributesKey[] {
            LABEL_DEFINITION
    };

    private static final TextAttributesKey[] LABEL_REFERENCE_KEYS = new TextAttributesKey[] {
            LABEL_REFERENCE
    };

    private static final TextAttributesKey[] EMPTY_KEYS = new TextAttributesKey[0];

    @NotNull
    @Override
    public Lexer getHighlightingLexer() {
        return new LatexLexerAdapter();
    }

    @NotNull
    @Override
    public TextAttributesKey[] getTokenHighlights(IElementType tokenType) {
        // Braces
        if (tokenType.equals(LatexTypes.OPEN_BRACE) ||
                tokenType.equals(LatexTypes.CLOSE_BRACE)) {
            return BRACES_KEYS;
        }
        // Brackets
        else if (tokenType.equals(LatexTypes.OPEN_BRACKET) ||
                tokenType.equals(LatexTypes.CLOSE_BRACKET)) {
            return BRACKET_KEYS;
        }
        // Comments
        else if (tokenType.equals(LatexTypes.COMMENT_TOKEN)) {
            return COMMENT_KEYS;
        }
        // Commands
        else if (COMMAND_TOKENS.contains(tokenType)) {
            return COMMAND_KEYS;
        }
        // Math environment
        else if (tokenType.equals(LatexTypes.INLINE_MATH_END)) {
            return INLINE_MATH_KEYS;
        }
        else if (tokenType.equals(LatexTypes.DISPLAY_MATH_START) ||
                tokenType.equals(LatexTypes.DISPLAY_MATH_END)) {
            return DISPLAY_MATH_KEYS;
        }
        // Star
        else if (tokenType.equals(LatexTypes.STAR)) {
            return STAR_KEYS;
        }
        // When no supported highlights is available
        else {
            return EMPTY_KEYS;
        }
    }
}

File: src/nl/hannahsten/texifyidea/highlighting/LatexSyntaxHighlighter.kt
package nl.hannahsten.texifyidea.highlighting

import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.tree.IElementType
import com.intellij.psi.tree.TokenSet
import nl.hannahsten.texifyidea.grammar.LatexLexerAdapter
import nl.hannahsten.texifyidea.psi.LatexTypes

/**
 * Syntax highlighting for lexer tokens (for composite elements, see [LatexAnnotator]).
 *
 * @author Hannah Schellekens, Sten Wessel
 */
class LatexSyntaxHighlighter : SyntaxHighlighterBase() {

    override fun getHighlightingLexer(): Lexer = LatexLexerAdapter()

    override fun getTokenHighlights(tokenType: IElementType): Array<out TextAttributesKey?> = if (tokenType == LatexTypes.OPEN_BRACE || tokenType == LatexTypes.CLOSE_BRACE) {
        BRACES_KEYS
    }
    else if (tokenType == LatexTypes.OPEN_BRACKET || tokenType == LatexTypes.CLOSE_BRACKET) {
        BRACKET_KEYS
    }
    else if (tokenType == LatexTypes.MAGIC_COMMENT_TOKEN) {
        MAGIC_COMMENT_KEYS
    }
    else if (tokenType == LatexTypes.COMMENT_TOKEN) {
        COMMENT_KEYS
    }
    else if (COMMAND_TOKENS.contains(tokenType)) {
        COMMAND_KEYS
    }
    else if (tokenType == LatexTypes.STAR) {
        STAR_KEYS
    }
    else {
        EMPTY_KEYS
    }

    companion object {

        /*
         * TextAttributesKeys
         */
        val BRACES = createKey("LATEX_BRACES", DefaultLanguageHighlighterColors.BRACES)
        val BRACKETS = createKey("LATEX_BRACKETS", DefaultLanguageHighlighterColors.BRACKETS)
        val OPTIONAL_PARAM = createKey("LATEX_OPTIONAL_PARAM", DefaultLanguageHighlighterColors.PARAMETER)
        val COMMAND = createKey("LATEX_COMMAND", DefaultLanguageHighlighterColors.KEYWORD)
        val USER_DEFINED_COMMAND = createKey("LATEX_USER_DEFINED_COMMAND", DefaultLanguageHighlighterColors.KEYWORD)
        val COMMAND_MATH_INLINE = createKey("LATEX_COMMAND_MATH_INLINE", DefaultLanguageHighlighterColors.VALID_STRING_ESCAPE)
        val COMMAND_MATH_DISPLAY = createKey("LATEX_COMMAND_MATH_DISPLAY", DefaultLanguageHighlighterColors.VALID_STRING_ESCAPE)
        val COMMENT = createKey("LATEX_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT)
        val MAGIC_COMMENT = createKey("LATEX_MAGIC_COMMENT", DefaultLanguageHighlighterColors.DOC_COMMENT)
        val INLINE_MATH = createKey("LATEX_INLINE_MATH", DefaultLanguageHighlighterColors.STRING)
        val DISPLAY_MATH = createKey("LATEX_DISPLAY_MATH", DefaultLanguageHighlighterColors.STRING)
        val STAR = createKey("LATEX_STAR", DefaultLanguageHighlighterColors.DOT)
        val SEPARATOR_EQUALS = createKey("SEPARATOR_EQUALS", DefaultLanguageHighlighterColors.IDENTIFIER)
        val LABEL_DEFINITION = createKey("LATEX_LABEL_DEFINITION", DefaultLanguageHighlighterColors.IDENTIFIER)
        val LABEL_REFERENCE = createKey("LATEX_LABEL_REFERENCE", DefaultLanguageHighlighterColors.IDENTIFIER)
        val BIBLIOGRAPHY_DEFINITION = createKey("LATEX_BIBLIOGRAPHY_DEFINITION", LABEL_DEFINITION)
        val BIBLIOGRAPHY_REFERENCE = createKey("LATEX_BIBLIOGRAPHY_REFERENCE", LABEL_REFERENCE)
        val STYLE_BOLD = createKey("LATEX_STYLE_BOLD", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_ITALIC = createKey("LATEX_STYLE_ITALIC", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_UNDERLINE = createKey("LATEX_STYLE_UNDERLINE", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_STRIKETHROUGH = createKey("LATEX_STYLE_STRIKETHROUGH", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_SMALL_CAPITALS = createKey("LATEX_STYLE_SMALL_CAPITALS", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_OVERLINE = createKey("LATEX_STYLE_OVERLINE", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_TYPEWRITER = createKey("LATEX_STYLE_TYPEWRITER", DefaultLanguageHighlighterColors.IDENTIFIER)
        val STYLE_SLANTED = createKey("LATEX_STYLE_SLANTED", DefaultLanguageHighlighterColors.IDENTIFIER)
        val MATH_NESTED_TEXT = createKey("LATEX_MATH_NESTED_TEXT", DefaultLanguageHighlighterColors.IDENTIFIER)

        private val COMMAND_TOKENS = TokenSet.create(
            LatexTypes.COMMAND_TOKEN,
            LatexTypes.COMMAND_IFNEXTCHAR,
            LatexTypes.BEGIN_TOKEN,
            LatexTypes.END_TOKEN,
            LatexTypes.BEGIN_PSEUDOCODE_BLOCK,
            LatexTypes.MIDDLE_PSEUDOCODE_BLOCK,
            LatexTypes.END_PSEUDOCODE_BLOCK,
            LatexTypes.START_IF,
            LatexTypes.ELSE,
            LatexTypes.END_IF,
            LatexTypes.LEFT,
            LatexTypes.RIGHT,
        )

        /*
         * TextAttributeKey[]s
         */
        private val BRACES_KEYS = keys(BRACES)
        private val BRACKET_KEYS = keys(BRACKETS)
        private val COMMAND_KEYS = keys(COMMAND)
        val USER_DEFINED_COMMAND_KEY = USER_DEFINED_COMMAND
        private val COMMENT_KEYS = keys(COMMENT)
        private val MAGIC_COMMENT_KEYS = keys(MAGIC_COMMENT)
        private val STAR_KEYS = keys(STAR)
        private val EMPTY_KEYS = arrayOfNulls<TextAttributesKey>(0)
        private fun createKey(externalName: String, defaultStyle: TextAttributesKey): TextAttributesKey = TextAttributesKey.createTextAttributesKey(externalName, defaultStyle)

        private fun keys(vararg keys: TextAttributesKey): Array<out TextAttributesKey> = keys
    }
}